src/libraries/Common/src/Interop/Linux/procfs/Interop.ProcFsStat.ParseMapModules.cs

```diff
@@ -0,0 +1,134 @@
+using System.Globalization;
+using System.IO;
+using System.Text;
+
+internal static partial class Interop
+{
+    internal static partial class procfs
+    {
+        private const string MapsFileName = "/maps";
+
+        private static string GetMapsFilePathForProcess(int pid)
+        {
+            return RootPath + pid.ToString(CultureInfo.InvariantCulture) + MapsFileName;
+        }
+
+        private static (long Start, int Size) TryParseAddressRange(string s, ref int start, ref int end)
+        {
+            int pos = s.IndexOf('-', start, end - start);
+            if (pos > 0)
+            {
+                if (long.TryParse(s.AsSpan(start, pos), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long startingAddress) &&
+                    long.TryParse(s.AsSpan(pos + 1, end - (pos + 1)), NumberStyles.HexNumber, CultureInfo.InvariantCulture, out long endingAddress))
+                {
+                    return (startingAddress, (int)(endingAddress - startingAddress));
+                }
+            }
+
+            return default;
+        }
+
+        private static bool HasReadAndExecFlags(string s, ref int start, ref int end)
+        {
+            bool sawRead = false, sawExec = false;
+            for (int i = start; i < end; i++)
+            {
+                if (s[i] == 'r')
+                    sawRead = true;
+                else if (s[i] == 'x')
+                    sawExec = true;
+            }
+
+            return sawRead & sawExec;
+        }
+
+        internal static ProcessModuleCollection? ParseMapsModules(int pid)
+        {
+            try
+            {
+                return ParseMapsModulesCore(File.ReadLines(GetMapsFilePathForProcess(pid)));
+            }
+            catch (IOException) { }
+            catch (UnauthorizedAccessException) { }
+
+            return null;
+        }
+
+        private static ProcessModuleCollection ParseMapsModulesCore(IEnumerable<string> lines)
+        {
+            Debug.Assert(lines != null);
+
+            ProcessModule? previous = null;
+            ProcessModuleCollection modules = new(capacity: 0);
+
+            foreach (string line in lines)
+            {
+                // Use a StringParser to avoid string.Split costs
+                var parser = new StringParser(line, separator: ' ', skipEmpty: true);
+
+                // Parse the address start and size
+                (long Start, int Size) addressRange = parser.ParseRaw(TryParseAddressRange);
+
+                if (addressRange == default)
+                {
+                    continue;
+                }
+
+                // Parse the permissions
+                bool hasReadAndExecFlags = parser.ParseRaw(HasReadAndExecFlags);
+
+                // Skip past the offset, dev, and inode fields
+                parser.MoveNext();
+                parser.MoveNext();
+                parser.MoveNext();
+
+                // Parse the pathname
+                if (!parser.MoveNext())
```

Comment:
In this case we may want to add a ProcessModule, e.g.:
```
55eb35f9c000-55eb35fc9000 r--p 00000000 fd:00 17965718                   /usr/bin/bash
55eb35fc9000-55eb3607f000 r-xp 0002d000 fd:00 17965718                   /usr/bin/bash
55eb3607f000-55eb360b6000 r--p 000e3000 fd:00 17965718                   /usr/bin/bash
55eb360b6000-55eb360ba000 r--p 00119000 fd:00 17965718                   /usr/bin/bash
55eb360ba000-55eb360c3000 rw-p 0011d000 fd:00 17965718                   /usr/bin/bash
55eb360c3000-55eb360cd000 rw-p 00000000 00:00 0 
```

I think the implementation needs to collect some state across the for-loop. Something like:
```
long moduleStart, moduleEnd;
bool moduleIsReadAndExec;
string modulePath;
...
for
{
    long lineStart, lineEnd;
    bool lineIsReadAndExec;
    string linePath;
    
    ...

    if (....)
    {
        modules.Add(new ProcessModule { ... });
    }
}
```

Reply:
We do not account for unnamed modules. PR is solely focused on fixing one issue as described in #45180.

Reply:
I mean: the `continue` on `!parser.MoveNext()` may lead to skipping a named module.

Reply:
> ```
> 55eb35f9c000-55eb35fc9000 r--p 00000000 fd:00 17965718                   /usr/bin/bash
> 55eb35fc9000-55eb3607f000 r-xp 0002d000 fd:00 17965718                   /usr/bin/bash
> 55eb3607f000-55eb360b6000 r--p 000e3000 fd:00 17965718                   /usr/bin/bash
> 55eb360b6000-55eb360ba000 r--p 00119000 fd:00 17965718                   /usr/bin/bash
> 55eb360ba000-55eb360c3000 rw-p 0011d000 fd:00 17965718                   /usr/bin/bash
> 55eb360c3000-55eb360cd000 rw-p 00000000 00:00 0
> ```
In this case, the last (swapper) row that has no name will be skipped by master and the PR, this condition has not changed. After this row, for the first named module, a new ProcessModule will be created on line 113.

Reply:
Can you tackle https://github.com/dotnet/runtime/issues/45482? When reviewing the code, I have trouble ignoring what is still missing to fix that.

Reply:
Fixed in 9fdc4eb.